src/shadow.rs

```diff
@@ -0,0 +1,224 @@
+        let Local{ ref pat, ref ty, ref init, id: _, span: _ } = **local;
+        if let &Some(ref t) = ty { check_ty(cx, t, bindings); }
+        named(pat).map(|name| if bindings.contains(&name) {
+            if let &Some(ref o) = init {
+                if in_external_macro(cx, o.span) { return; }
+                check_expr(cx, o, bindings);
+                bindings.push(name);
+                lint_shadow(cx, name, decl.span, pat.span, o);
+            }
+        });
+        add(bindings, pat);
+        if let &Some(ref o) = init {
+            check_expr(cx, o, bindings)
+        }
+    }
+}
+
+fn lint_shadow(cx: &Context, name: Name, span: Span, lspan: Span, init: &Expr) {
+    if is_self_shadow(name, init) {
+        span_lint(cx, SHADOW_SAME, span, &format!(
+            "{} is shadowed by itself in {}",
+            snippet(cx, lspan, "_"),
+            snippet(cx, init.span, "..")));
+    } else {
+        if contains_self(name, init) {
+            span_lint(cx, SHADOW_REUSE, span, &format!(
+                "{} is shadowed by {} which reuses the original value",
+                snippet(cx, lspan, "_"),
+                snippet(cx, init.span, "..")));
+        } else {
+            span_lint(cx, SHADOW_FOREIGN, span, &format!(
+                "{} is shadowed by {} in this declaration",
+                snippet(cx, lspan, "_"),
+                snippet(cx, init.span, "..")));
+        }
+    }
+}
+
+fn check_expr(cx: &Context, expr: &Expr, bindings: &mut Vec<Name>) {
+    if in_external_macro(cx, expr.span) { return; }
+    match expr.node {
+        ExprUnary(_, ref e) | ExprParen(ref e) | ExprField(ref e, _) |
+        ExprTupField(ref e, _) | ExprAddrOf(_, ref e) | ExprBox(None, ref e)
+            => { check_expr(cx, e, bindings) },
+        ExprBox(Some(ref place), ref e) => {
+            check_expr(cx, place, bindings); check_expr(cx, e, bindings) }
+        ExprBlock(ref block) | ExprLoop(ref block, _) =>
+            { check_block(cx, block, bindings) },
+        ExprVec(ref v) | ExprTup(ref v) =>
+            for ref e in v { check_expr(cx, e, bindings) },
+        ExprIf(ref cond, ref then, ref otherwise) => {
+            check_expr(cx, cond, bindings);
+            check_block(cx, then, bindings);
+            if let &Some(ref o) = otherwise { check_expr(cx, o, bindings); }
+        },
+        ExprIfLet(ref pat, ref e, ref block, ref otherwise) => {
+            check_expr(cx, e, bindings);
+            let len = bindings.len();
+            add(bindings, pat);
+            check_block(cx, block, bindings);
+            if let &Some(ref o) = otherwise { check_expr(cx, o, bindings); }
+            bindings.truncate(len);
+        },
+        ExprWhile(ref cond, ref block, _) => {
+            check_expr(cx, cond, bindings);
+            check_block(cx, block, bindings);
+        },
+        ExprWhileLet(ref pat, ref e, ref block, _) |
+        ExprForLoop(ref pat, ref e, ref block, _) => {
```

Comment:
`ExprForLoop`, `ExprIfLet` and `ExprWhileLet` will not be in the AST anymore at this stage.